explorer/interpreter/interpreter.cpp

```diff
@@ -1088,21 +1088,49 @@ auto Interpreter::CallFunction(const CallExpression& call,
 }
 
 auto Interpreter::StepInstantiateType() -> ErrorOr<Success> {
-  const Action& act = todo_.CurrentAction();
-  const Nonnull<const Value*> type = cast<TypeInstantiationAction>(act).type();
-  SourceLocation source_loc = cast<TypeInstantiationAction>(act).source_loc();
+  const auto& act = cast<TypeInstantiationAction>(todo_.CurrentAction());
+  const Nonnull<const Value*> type = act.type();
+  SourceLocation source_loc = act.source_loc();
+  bool base_type = act.base_type();
 
   switch (type->kind()) {
     case Value::Kind::NominalClassType: {
       const auto& class_type = cast<NominalClassType>(*type);
       std::optional<Nonnull<const NominalClassType*>> base = class_type.base();
       if (act.pos() == 0 && base.has_value()) {
         return todo_.Spawn(std::make_unique<TypeInstantiationAction>(
-            base.value(), source_loc));
+            base.value(), source_loc, /*base_type=*/true));
       } else {
         if (base.has_value()) {
           base = cast<NominalClassType>(act.results().back());
         }
+        if (!base_type && class_type.declaration().extensibility() ==
+                              ClassExtensibility::Abstract) {
+          return ProgramError(source_loc)
+                 << "Cannot instantiate abstract class "
+                 << class_type.declaration().name();
+        }
+
+        if (!base_type && class_type.declaration().extensibility() ==
+                              ClassExtensibility::Base) {
+          const auto& class_vtable = class_type.vtable();
+          auto abstract_method_it = std::find_if(
+              class_vtable.begin(), class_vtable.end(), [](const auto& vt) {
+                const auto* const fun = vt.getValue().first;
+                return fun->is_method() &&
+                       fun->virt_override() == VirtualOverride::Abstract;
+              });
+
+          if (abstract_method_it != class_vtable.end()) {
+            auto fun_name = GetName(*abstract_method_it->getValue().first);
+            CARBON_CHECK(fun_name.has_value());
+            return ProgramError(source_loc)
+                   << "Cannot instantiate base class `"
+                   << class_type.declaration().name() << "`: abstract method `"
+                   << *fun_name << "` should be implemented.";
+          }
```

Comment:
addressing https://github.com/carbon-language/carbon-lang/pull/2528#discussion_r1095393751

I have replaced for loop with find_if. So this is more readable now. 
And we can't unify the `CARBON_CHECK...` because the error messages are slightly different. 
One option is to just move `find_if()` part to common function. But, I am not this is useful.